File: src/shared/ipc-contracts/device-catalog.ts
/**
 * Wire contract for the official-catalog commands (story 2-6, Phase C).
 * Mirror of `src-tauri/src/ipc/dto/device_catalog.rs`.
 *
 * `CatalogStatusDto` is returned by `get_official_catalog_status` and
 * `refresh_official_catalog`. `ImportOfficialCatalogOutcome` is returned by
 * `import_official_catalog` — a tagged enum, because a cancelled file dialog
 * is a normal (non-error) outcome, exactly like the export flow.
 */

export interface CatalogStatusDto {
  /** Number of official titles currently cached locally. */
  count: number;
}

export type ImportOfficialCatalogOutcome =
  | { kind: "cancelled" }
  | { kind: "imported"; count: number };

/** One cached cover served as a self-contained `data:` URL (read from the
 *  local cache, no network). Mirror of `PackCoverDto`. */
export interface PackCoverDto {
  dataUrl: string;
}

function hasOnlyAllowedKeys(
  value: Record<string, unknown>,
  allowed: ReadonlySet<string> | undefined,
): boolean {
  if (!allowed) return false;
  for (const k of Object.keys(value)) {
    if (!allowed.has(k)) return false;
  }
  return true;
}

const STATUS_KEYS: ReadonlySet<string> = new Set(["count"]);

function isCount(value: unknown): value is number {
  return typeof value === "number" && Number.isInteger(value) && value >= 0;
}

/** Runtime guard for a `CatalogStatusDto`. */
export function isCatalogStatusDto(value: unknown): value is CatalogStatusDto {
  if (typeof value !== "object" || value === null) return false;
  const c = value as Record<string, unknown>;
  if (!hasOnlyAllowedKeys(c, STATUS_KEYS)) return false;
  return isCount(c.count);
}

const IMPORT_KEYS: Record<string, ReadonlySet<string>> = {
  cancelled: new Set(["kind"]),
  imported: new Set(["kind", "count"]),
};

/** Runtime guard for an `ImportOfficialCatalogOutcome`. */
export function isImportOfficialCatalogOutcome(
  value: unknown,
): value is ImportOfficialCatalogOutcome {
  if (typeof value !== "object" || value === null) return false;
  const c = value as Record<string, unknown>;
  if (typeof c.kind !== "string") return false;
  if (!hasOnlyAllowedKeys(c, IMPORT_KEYS[c.kind])) return false;
  switch (c.kind) {
    case "cancelled":
      return true;
    case "imported":
      return isCount(c.count);
    default:
      return false;
  }
}

const COVER_KEYS: ReadonlySet<string> = new Set(["dataUrl"]);

/** Runtime guard for a `PackCoverDto` — a non-empty `data:` URL. */
export function isPackCoverDto(value: unknown): value is PackCoverDto {
  if (typeof value !== "object" || value === null) return false;
  const c = value as Record<string, unknown>;
  if (!hasOnlyAllowedKeys(c, COVER_KEYS)) return false;
  return typeof c.dataUrl === "string" && c.dataUrl.startsWith("data:");
}

File: src-tauri/src/ipc/dto/device_catalog.rs
use serde::Serialize;

/// Current state of the official-catalog cache. Mirror of
/// `src/shared/ipc-contracts/device-catalog.ts`.
#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct CatalogStatusDto {
    /// Number of official titles currently cached locally.
    pub count: u32,
}

impl CatalogStatusDto {
    pub fn new(count: u32) -> Self {
        Self { count }
    }
}

/// Outcome of `import_official_catalog`. A cancelled file dialog is NOT an
/// error — it resolves with `{ kind: "cancelled" }` so the UI returns to
/// idle silently, mirroring the export flow.
#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
#[serde(tag = "kind", rename_all = "camelCase")]
pub enum ImportOfficialCatalogOutcomeDto {
    Cancelled,
    #[serde(rename_all = "camelCase")]
    Imported {
        count: u32,
    },
}

/// One cached cover, returned by `read_pack_cover` as a self-contained
/// `data:` URL the webview can render directly — read from the LOCAL cache,
/// no network. Mirror of `src/shared/ipc-contracts/device-catalog.ts`.
#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct PackCoverDto {
    pub data_url: String,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn status_serializes_in_camel_case() {
        let v = serde_json::to_value(CatalogStatusDto::new(42)).expect("ser");
        assert_eq!(v, serde_json::json!({ "count": 42 }));
    }

    #[test]
    fn import_outcome_cancelled_serializes_with_kind() {
        let v = serde_json::to_value(ImportOfficialCatalogOutcomeDto::Cancelled).expect("ser");
        assert_eq!(v, serde_json::json!({ "kind": "cancelled" }));
    }

    #[test]
    fn import_outcome_imported_serializes_count() {
        let v = serde_json::to_value(ImportOfficialCatalogOutcomeDto::Imported { count: 7 })
            .expect("ser");
        assert_eq!(v, serde_json::json!({ "kind": "imported", "count": 7 }));
    }
}
